source/simulator/lateral_averaging.cc

```diff
@@ -448,17 +250,279 @@ namespace aspect
         std::vector<Tensor<1,dim> > temperature_gradients;
         std::vector<double> temperature_values;
     };
+  }
+
+  template <int dim>
+  void LateralAveraging<dim>::compute_lateral_averages(std::vector<std_cxx11::unique_ptr<FunctorBase<dim> > > &functors,
+                                                       std::vector<std::vector<double> > &values) const
+  {
+    Assert (values.size() > 0,
+            ExcMessage ("To call this function, you need to request a positive "
+                        "number of properties to compute."));
+    Assert (values[0].size() > 0,
+            ExcMessage ("To call this function, you need to request a positive "
+                        "number of depth slices."));
+    Assert (functors.size() == values.size(),
+            ExcMessage ("To call this function, you need supply as many data "
+                        "vectors as properties you want to compute."));
+
+    const unsigned int n_properties = values.size();
+    const unsigned int n_slices = values[0].size();
+    for (unsigned int i=0; i < n_properties; ++i)
+      Assert (values[i].size() == n_slices,
+              ExcMessage ("All properties need to have the same number of depth slices."));
+
+    std::vector<double> volume(n_slices);
+
+    // this yields 10^dim quadrature points evenly distributed in the interior of the cell.
+    // We avoid points on the faces, as they would be counted more than once.
+    const QIterated<dim> quadrature_formula (QMidpoint<1>(),
+                                             10);
+    const unsigned int n_q_points = quadrature_formula.size();
+    const double max_depth = this->get_geometry_model().maximal_depth();
+
+    FEValues<dim> fe_values (this->get_mapping(),
+                             this->get_fe(),
+                             quadrature_formula,
+                             update_values | update_gradients | update_quadrature_points | update_JxW_values);
+
+    std::vector<std::vector<double> > composition_values (this->n_compositional_fields(),std::vector<double> (n_q_points));
+
+    std::vector<std::vector<double> > output_values(n_properties,std::vector<double>(quadrature_formula.size()));
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = this->get_dof_handler().begin_active(),
+    endc = this->get_dof_handler().end();
+
+
+    MaterialModel::MaterialModelInputs<dim> in(n_q_points,
+                                               this->n_compositional_fields());
+    MaterialModel::MaterialModelOutputs<dim> out(n_q_points,
+                                                 this->n_compositional_fields());
+
+    bool functors_need_material_output = false;
+    for (unsigned int i=0; i<n_properties; ++i)
+      {
+        functors[i]->setup(quadrature_formula.size());
+        if (functors[i]->need_material_properties())
+          functors_need_material_output = true;
```

Comment:
I look forward to the day where we can use `std::reduce` or `std::transform_reduce` for this kind of thing -- but that's only in C++17 :-(
http://en.cppreference.com/w/cpp/algorithm/reduce
http://en.cppreference.com/w/cpp/algorithm/transform_reduce